### PR TITLE
3586 add first production test experiments

### PR DESCRIPTION
**Pull request by kpethtel**
**Story card:** [ch3586](https://app.clubhouse.io/simpledotorg/story/3586/test-in-sandbox)

## Because

It's time to start running experiments with real patients.  We intent to run the first experiment with ~300 patients so that we can monitor and prevent any sort of errors or failed messages from impacting the large patient set in India production.

## How to Run
* [ ] read over the HOWTO for an overview on the general process - see the [doc here added in this PR](https://github.com/simpledotorg/simple-server/blob/3586-add-first-production-test-experiments/doc/howto/first_ab_experiment.md).
* [ ] merge this PR
* [ ] verify the PR has deployed to all environments successfully
* [ ] now you can run through the steps in the HOWTO doc, going off the version of it in master
* [ ] refine or update anything that came for the docs for future runs, and also please leave context in the linked Clubhouse story

## Monitoring: 
  - Unfortunately, I have not yet changed the TwilioApiService back to reporting errors that have specified codes to Sentry, so some errors may appear in the Datadog logs. You should be able to find them by searching for the classname `TwilioApiService`. 
  - Errors with unknown causes should still show up in Sentry with the message: "Error while calling Twilio API"

### CODE REVIEW

File: db/data_schema.rb [hunk 1]
```diff
@@ -1,2 +1,2 @@
 # encoding: UTF-8
-DataMigrate::Data.define(version: 20210708133536)
+DataMigrate::Data.define(version: 20210725213406)
```
Comment (kpethtel) [inline db/data_schema.rb:2]: I assume this should be included, similar to how db/schema works.

File: lib/seed/production_experiment_test_seeder.rb [hunk 2]
```diff
@@ -1,24 +0,0 @@
-module Seed
```
Comment (kpethtel) [inline lib/seed/production_experiment_test_seeder.rb:1]: This didn't seem to have any lasting value so I removed it.

**Comment by rsanheim** [comment]
> `ExperimentControlService.send(:current_patient_candidates, "July 28, 2021".to_date, "July 30, 2021".to_date).count`

This may operate in UTC in a standard Rails console if we aren't careful, which I don't think is what we want. The operator probably should run `Time.zone = Period::REPORTING_TIME_ZONE` before doing any of this.

>  use that number as the last argument when starting the current patient experiment (the numerical arguments are 1 day from now, 3 days from now, and 3%): `ExperimentControlService.start_current_patient_experiment("Small Current Patient July 2021", 1, 3, 3)`

We should refactor the runner methods to have named args, then this would become much more clear.

> runner "AppointmentNotification::ScheduleExperimentReminders.perform_now"

Is this meant to be done from the rails console, or from a shell? If its a shell, it should be `rails runner` I think.

I would expect to see something about keeping an eye on datadog logs, Sentry, and perhaps the Sidekiq queues in here as well.

**Comment by kpethtel** [comment]
> > `ExperimentControlService.send(:current_patient_candidates, "July 28, 2021".to_date, "July 30, 2021".to_date).count`
> 
> This may operate in UTC in a standard Rails console if we aren't careful, which I don't think is what we want. The operator probably should run `Time.zone = Period::REPORTING_TIME_ZONE` before doing any of this.

I'll make this change. 

> > use that number as the last argument when starting the current patient experiment (the numerical arguments are 1 day from now, 3 days from now, and 3%): `ExperimentControlService.start_current_patient_experiment("Small Current Patient July 2021", 1, 3, 3)`
> 
> We should refactor the runner methods to have named args, then this would become much more clear.

I'll agree with that. Not sure I want to make those changes with so little time available.

> > runner "AppointmentNotification::ScheduleExperimentReminders.perform_now"
> 
> Is this meant to be done from the rails console, or from a shell? If its a shell, it should be `rails runner` I think.
> 

I was trying not to be too prescriptive about it, but perhaps I should be. 

> I would expect to see something about keeping an eye on datadog logs, Sentry, and perhaps the Sidekiq queues in here as well.

I'll make a comment to that effect, but I haven't really have specifics for what to look for.

File: doc/howto/first_ab_experiment.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# How to run the initial AB Experiment in India Production 
+
+### Date: July 2021
+### References
+* [clubhouse story ch3586](https://app.clubhouse.io/simpledotorg/story/3586/test-in-sandbox)
+* [GitHub PR](https://github.com/simpledotorg/simple-server/pull/2752)
```
Comment (rsanheim) [inline doc/howto/first_ab_experiment.md:6]: There is probably more we can link here, but this is a good start for context.

File: doc/howto/first_ab_experiment.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+---
+
+
+*Note: these instructions are only valid for Tuesday July 27 before 4pm IST (the end of the messaging window).*
+
+1. Turn on `experiment` and `whatsapp_appointment_reminders` Flipper feature flags
+2. Merge https://github.com/simpledotorg/simple-server/pull/2752 - it includes a migration that will create two new experiments with three treatment groups each. To verify:
+
+```ruby
```
Comment (rsanheim) [inline doc/howto/first_ab_experiment.md:16]: for syntax highlighting for readability

File: doc/howto/first_ab_experiment.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+experiments = Experiment.where(name: "Small Current Patient July 2021")
+experiments.count # should return 2
+experiments.map { |e| e.treatment_groups.count } # should return 3
```
Comment (rsanheim) [inline doc/howto/first_ab_experiment.md:19]: Including the exact code that the operator can run to verify.

File: doc/howto/first_ab_experiment.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+3. Create the current patient experiment, based on how many patients are eligible. We want to use approximately 300 patients and this particular job uses percentages. So if 10,000 patients are eligible, we would want to use 3% (300/10,000). 
+**NOTE:** change the percentage below so that the experiment enrolls ~300 patients
+
+  ```ruby
+  Time.zone = Period::REPORTING_TIME_ZONE
+
+  candidates = ExperimentControlService.send(:current_patient_candidates, "July 28, 2021".to_date, "July 30, 2021".to_date).count
+  # => 10000 ...assume there are 10,000 eligible patients returned...
+  percentage = 300 / candidates # results in 0.03
```
Comment (rsanheim) [inline doc/howto/first_ab_experiment.md:30]: Trying to ensure we really end up with 300 patients in the actual experiment...sometimes 'this is for example use only!' text gets ignored =)

File: doc/howto/first_ab_experiment.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+  ExperimentControlService.start_current_patient_experiment("Small Current Patient July 2021", 1, 3, percentage) # the final argument is the percentage of patients to enroll
+
+  # Verification
+  experiment = Experiment.find_by!(name: "Small Current Patient July 2021")
+  experiment.notifications.count
+  # => should return around 400 notifications
```
Comment (rsanheim) [inline doc/howto/first_ab_experiment.md:36]: by commenting out notes like this, it means the operator can copy paste blocks of commands and it won't execute the docs

File: doc/howto/first_ab_experiment.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+4. Launch the job to schedule the first wave of notifications from a rails console.
+The changes may not be immediate due to the nature of background jobs, but this should result in the notifications that are due today being changed to status "sent". It should also create communications for those notifications.
+
+
+```ruby
+AppointmentNotification::ScheduleExperimentReminders.perform_now
+# Run the following after the job has finished running (check Sidekiq) to verify:
+ns = Notification.where(remind_on: Date.current)
+ns.pluck(:status).uniq  # (should all equal "sent")
+ns.map {|n| n.communications.count}.uniq # (expect 1 and maybe 2; may still include 0 due to twilio errors)
```
Comment (rsanheim) [inline doc/howto/first_ab_experiment.md:48]: Not the best confirmation steps here, but I'm sure we can refine this as we get better at this process

File: doc/howto/first_ab_experiment.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+```
+
+### Troubleshooting
+
+* Monitor Sentry and Datadog logs for any errors or anomalies while running
```
Comment (rsanheim) [inline doc/howto/first_ab_experiment.md:54]: Would be good to link up searches or dashboards here, at least for Datadog since it can be overwhelming

**Review by rsanheim — COMMENTED** [review]
Notes on the new docs

**Comment by rsanheim** [comment]
Hey @kpethtel @prabhanshuguptagit - I've moved the docs to a HOWTO markdown file and tried to make it easier to be run in production. Please do update the doc in this PR with any changes to make things more clear, and we can iterate on this as we go to get the process more streamlined.  Also, if any errors do come up when running things, please drop the the full logs into this PR or the linked Clubhouse story. Thanks!

**Review by prabhanshuguptagit — COMMENTED** [review]
Minor fixes to README for next time

**Pull request merged**